File: python/src/palimpsest/gate_b/artifacts.py
from __future__ import annotations

import os
from pathlib import Path
from typing import Any

from palimpsest.contracts import canonical_json_bytes, sha256_hex


def artifact_reference(path: Path, artifact_type: str) -> dict[str, Any]:
    content = path.read_bytes()
    return {
        "artifactType": artifact_type,
        "byteLength": len(content),
        "sha256": sha256_hex(content),
    }


def promote_file(path: Path, artifact_type: str) -> dict[str, Any]:
    content = path.read_bytes()
    reference = {
        "artifactType": artifact_type,
        "byteLength": len(content),
        "sha256": sha256_hex(content),
    }
    root = Path("artifacts/gate-b/by-digest")
    root.mkdir(parents=True, exist_ok=True)
    destination = root / reference["sha256"]
    if destination.exists():
        if destination.read_bytes() != content:
            raise RuntimeError(f"Digest store collision at {destination}.")
    else:
        staging = root / f".{reference['sha256']}.{os.getpid()}.tmp"
        staging.write_bytes(content)
        try:
            staging.replace(destination)
        except OSError:
            if not destination.exists() or destination.read_bytes() != content:
                raise
            staging.unlink(missing_ok=True)
    return reference


def write_canonical(path: Path, value: Any) -> dict[str, Any]:
    content = canonical_json_bytes(value)
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_bytes(content)
    return {
        "artifactType": path.stem.replace("_", "-"),
        "byteLength": len(content),
        "sha256": sha256_hex(content),
    }


def write_text(path: Path, value: str, artifact_type: str) -> dict[str, Any]:
    content = value.encode("utf-8")
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_bytes(content)
    return {
        "artifactType": artifact_type,
        "byteLength": len(content),
        "sha256": sha256_hex(content),
    }

File: tools/gate-a/artifacts.ts
import { mkdir, readFile, readdir, rename, rm, stat, writeFile } from "node:fs/promises";
import { dirname, join, relative, sep } from "node:path";

import { canonicalArchiveBytes, canonicalJsonBytes, sha256Hex } from "@palimpsest/contracts";

export interface ArtifactReference {
  artifactType: string;
  byteLength: number;
  sha256: string;
}

export async function writeCanonicalJson(path: string, value: unknown): Promise<Buffer> {
  const bytes = canonicalJsonBytes(value);
  await mkdir(dirname(path), { recursive: true });
  await writeFile(path, bytes);
  return bytes;
}

export async function referenceFile(
  path: string,
  artifactType: string,
): Promise<ArtifactReference> {
  const bytes = await readFile(path);
  return {
    artifactType,
    byteLength: bytes.length,
    sha256: sha256Hex(bytes),
  };
}

async function listFiles(root: string, current = root): Promise<string[]> {
  const paths: string[] = [];
  for (const name of (await readdir(current)).sort()) {
    if (
      name === ".DS_Store" ||
      name === ".cache" ||
      name === ".pytest_cache" ||
      name === ".ruff_cache" ||
      name === "__pycache__" ||
      name.endsWith(".pyc")
    ) {
      continue;
    }
    const path = join(current, name);
    const metadata = await stat(path);
    if (metadata.isDirectory()) {
      paths.push(...(await listFiles(root, path)));
    } else if (metadata.isFile()) {
      paths.push(relative(root, path).split(sep).join("/"));
    }
  }
  return paths;
}

export async function referenceBundle(
  artifactType: string,
  roots: string[],
): Promise<ArtifactReference> {
  const entries = [];
  for (const root of [...roots].sort()) {
    const metadata = await stat(root);
    if (metadata.isFile()) {
      entries.push({
        contentBase64: (await readFile(root)).toString("base64"),
        kind: "file" as const,
        path: root.split(sep).join("/"),
      });
      continue;
    }
    for (const path of await listFiles(root)) {
      entries.push({
        contentBase64: (await readFile(join(root, path))).toString("base64"),
        kind: "file" as const,
        path: join(root, path).split(sep).join("/"),
      });
    }
  }
  const bytes = canonicalArchiveBytes({
    contractId: "canonical-archive",
    entries,
    schemaVersion: 1,
  });
  return {
    artifactType,
    byteLength: bytes.length,
    sha256: sha256Hex(bytes),
  };
}

export async function promoteBytes(
  bytes: Buffer,
  artifactType: string,
  root = "artifacts/gate-a/by-digest",
): Promise<ArtifactReference> {
  const sha256 = sha256Hex(bytes);
  const destination = join(root, sha256);
  await mkdir(root, { recursive: true });
  const existing = await readFile(destination).catch(() => undefined);
  if (existing) {
    if (!existing.equals(bytes)) {
      throw new Error(`Digest store collision at ${destination}.`);
    }
  } else {
    const staging = join(root, `.${sha256}.${process.pid}.tmp`);
    await writeFile(staging, bytes);
    try {
      await rename(staging, destination);
    } catch (error) {
      const raced = await readFile(destination).catch(() => undefined);
      if (!raced?.equals(bytes)) {
        throw error;
      }
      await rm(staging, { force: true });
    }
  }
  return { artifactType, byteLength: bytes.length, sha256 };
}
